### DISCUSSION TITLE
A link-prediction question from a green hand: why my model seems not able to learn anything?

### DISCUSSION THREAD

**Discussion by danphnis** [pyg-team/pytorch_geometric#4919] [Q&A] [answered]
I was trying to try this wonderful package for KG on a simple case:
```
import numpy as np
cases = np.arange(200)
chips = np.arange(200)
versions = np.arange(20)
in_versions = {}
for c in chips:
    v = np.random.randint(20)
    in_versions[c] = v
    
chips_fail_case = {}
version_fail_case = {}
for chip in chips:
    for case in cases:
        if np.random.randint(8) == 0: 
            chips_fail_case[(chip,case)] = 1
            v = in_versions[chip]
            version_fail_case[(v,case)] = 1
```
Suppose that I have 200 "cases" to test the 200 "chips" from 20 "versions" respectively. 
Every "chip" is linked to its corresponding "version" exclusively. 
If the "chip" fails a test "case", then it is linked to that "case".
And if any "chip" from a certain "version" fails a test "case", the "version" is also linked to that "case".
Finally, I want to predict the links between “version” and "case". I though it would be a simple task, since the "link" can be predict from whether any "chips" exists linked to both the "version" and the "case" which is set to be known.

I have built the HeteroGraphic as follow:
```
ata = HeteroData()
data['case'].x = torch.eye(len(cases))
data['chip'].x = torch.eye(len(chips))
data['version'].x = torch.eye(len(versions))

links = np.array([(c, in_versions[c])for c in in_versions])
data['chip','in','version'].edge_index = torch.from_numpy(links.T).to(torch.long)

data['version','fail','case'].edge_index = torch.from_numpy(np.array(list(version_fail_case.keys())).T).to(torch.long)
data['version','fail','case'].edge_label = torch.from_numpy(np.array(list(version_fail_case.values())).reshape(-1,1)).to(torch.long)

data['chip','fail','case'].edge_index = torch.from_numpy(np.array(list(chips_fail_case.keys())).T).to(torch.long)
data['chip','fail','case'].edge_label = torch.from_numpy(np.array(list(chips_fail_case.values())).reshape(-1,1)).to(torch.long)

device = 'cuda' if torch.cuda.is_available() else 'cpu'

data = T.ToUndirected()(data)
del data['case', 'rev_fail', 'version'].edge_label
```

And I followed the example for link predicion as below:
```
train_data, val_data, test_data = T.RandomLinkSplit(
    num_val=0.1,
    num_test=0.1,
    neg_sampling_ratio=1.0,
    edge_types=[('version', 'fail', 'case')],
    rev_edge_types=[('case', 'rev_fail', 'version')],
)(data)

class GNNEncoder(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels):
        super().__init__()
        self.convs = torch.nn.ModuleList()
        for _ in range(1):
            conv = HeteroConv({
                edge_type: SAGEConv((-1, -1), hidden_channels,)
                for edge_type in data.edge_types
            },aggr='sum')
            self.convs.append(conv)

    def forward(self, x_dict, edge_index_dict):
        for conv in self.convs:
            x_dict = conv(x_dict, edge_index_dict)
            x_dict = {key: x for key, x in x_dict.items()}
        return x_dict
        
class EdgeDecoder(torch.nn.Module):
    def __init__(self, hidden_channels):
        super().__init__()
        self.lin1 = Linear(2*hidden_channels, hidden_channels)
        self.lin2 = Linear(hidden_channels, 1)

    def forward(self, z_dict, edge_label_index):
        row, col = edge_label_index
        z = torch.cat([z_dict['version'][row], z_dict['case'][col]], dim=-1)
#        z = z_dict['version'][row] - z_dict['case'][col]
        z = self.lin1(z).relu()
        z = self.lin2(z)#.sigmoid()
        return z.view(-1)
    
class Model(torch.nn.Module):
    def __init__(self, hidden_channels):
        super().__init__()
        self.encoder = GNNEncoder(hidden_channels, hidden_channels)
        self.decoder = EdgeDecoder(hidden_channels)

    def forward(self, x_dict, edge_index_dict, edge_label_index):
        z_dict = self.encoder(x_dict, edge_index_dict)
        return self.decoder(z_dict, edge_label_index)

model = Model(hidden_channels=200).to(device)
model.encoder(train_data.x_dict, train_data.edge_index_dict)

optimizer = torch.optim.Adam(model.parameters(), lr=0.01)


def train():
    model.train()
    optimizer.zero_grad()
    pred = model(train_data.x_dict, train_data.edge_index_dict,
                 train_data['version', 'case'].edge_label_index)
    target = train_data['version', 'case'].edge_label.float()/2
    loss = F.mse_loss(pred, target)
    loss.backward()
    optimizer.step()
    return float(loss)


@torch.no_grad()
def test(data):
    model.eval()
    pred = model(data.x_dict, data.edge_index_dict,
                 data['version', 'case'].edge_label_index)
    target = data['version', 'case'].edge_label.float()/2
    rmse = F.mse_loss(pred, target.squeeze())
    return float(rmse)


for epoch in range(1, 200):
    loss = train()
    train_rmse = test(train_data)
    val_rmse = test(val_data)
    test_rmse = test(test_data)
    print(f'Epoch: {epoch:03d}, Loss: {loss:.4f}, Train: {train_rmse:.4f}, '
          f'Val: {val_rmse:.4f}, Test: {test_rmse:.4f}')
```

However, it turns out that the model seems not able to actually learn any experience. The trainning loss stop at 0.1587. And the model  just predict the existance between any "case" nodes and "version" nodes very similarly (the value of pred varies very little). 

Could you please often me some suggestions? I will be extremely gratefull.

**Comment by danphnis** [comment, marked as answer]
This problem has got me puzzled for the whole afternoon. However, I think I have fixed it just a few minutes after I posted it here. I didn't notice there was a warning that:

UserWarning: Using a target size (torch.Size([4600, 1])) that is different to the input size (torch.Size([4600])). This will likely lead to incorrect results due to broadcasting. Please ensure they have the same size.

Therefore, in the function "train", I squeezed the tensor "target" like:

```
def train():
    model.train()
    optimizer.zero_grad()
    pred = model(train_data.x_dict, train_data.edge_index_dict,
                 train_data['version', 'case'].edge_label_index)
    target = train_data['version', 'case'].edge_label.float()/2
    loss = F.mse_loss(pred, target.squeeze())
    loss.backward()
    optimizer.step()
    return float(loss)
```

Thanks anyway!

**Reply by rusty1s** [reply]
Super, glad that it is resolved.
